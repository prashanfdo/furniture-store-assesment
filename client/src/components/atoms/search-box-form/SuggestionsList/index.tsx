import React from "react";
import { Link } from "react-router-dom";

type SuggestionsListProps = {
  suggestions: {
    name: string;
    results: number;
    image: string;
  }[];
  showSuggestions: boolean;
};

// TODO: Link to category page
const SuggestionsList: React.FC<SuggestionsListProps> = ({ suggestions, showSuggestions }: SuggestionsListProps) => {
  if (!showSuggestions || suggestions.length === 0) {
    return null;
  }

  return (
    <ul
      data-testid="suggestion-list"
      className="fixed left-0 hidden w-full pb-2 mt-2 bg-white rounded-md suggestion-list md:absolute top-16 md:top-full drop-shadow-md md:drop-shadow-lg group-focus-within:block"
    >
      {suggestions.map((suggestion) => (
        <li key={suggestion.name}>
          <Link className="block px-4 py-2 decoration-transparent" to="#">
            {suggestion.name}
          </Link>
        </li>
      ))}
    </ul>
  );
};

export default SuggestionsList;
