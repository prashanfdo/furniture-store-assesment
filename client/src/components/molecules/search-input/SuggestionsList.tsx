import React from "react";
import { Link } from "react-router-dom";

type SuggestionsListProps = {
  suggestions: {
    name: string;
    results: number;
    image: string;
  }[];
  showSuggestions: boolean;
};

// TODO: Link to category page
export const SuggestionsList: React.FC<SuggestionsListProps> = ({
  suggestions,
  showSuggestions,
}: SuggestionsListProps) => {
  if (!showSuggestions || suggestions.length === 0) {
    return null;
  }

  return (
    <ul className="suggestion-list md:absolute mt-2 bg-white w-full top-16 md:top-full rounded-md drop-shadow-md md:drop-shadow-lg pb-2 fixed left-0">
      {suggestions.map((suggestion) => (
        <li key={suggestion.name}>
          <Link className="block decoration-transparent py-2 px-4" to="#">
            {suggestion.name}
          </Link>
        </li>
      ))}
    </ul>
  );
};
