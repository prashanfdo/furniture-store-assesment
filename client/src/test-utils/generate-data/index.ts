export * from "./article";
export * from "./category";
export * from "./search";
