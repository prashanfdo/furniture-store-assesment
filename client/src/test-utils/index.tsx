import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
