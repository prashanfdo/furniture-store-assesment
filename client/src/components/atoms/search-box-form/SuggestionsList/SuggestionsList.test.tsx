import user from "@testing-library/user-event";
import { render, screen } from "test-utils";
import SuggestionsList from "./index";

const dataSuggestions: React.ComponentProps<typeof SuggestionsList>["suggestions"] = [
  {
    name: "test 1",
    results: 1,
    image: "",
  },
  {
    name: "test 2",
    results: 2,
    image: "",
  },
  {
    name: "test 3",
    results: 3,
    image: "",
  },
];

describe("SuggestionsList", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const { container } = render(<SuggestionsList suggestions={dataSuggestions} showSuggestions={true} />);
    expect(container).toMatchSnapshot();
  });

  it("should render links", () => {
    render(<SuggestionsList suggestions={dataSuggestions} showSuggestions={true} />);
    expect(screen.getAllByRole("link")).toHaveLength(dataSuggestions.length);
  });

  it("should not visible with no suggestions", () => {
    render(<SuggestionsList suggestions={[]} showSuggestions={true} />);
    expect(screen.queryByTestId("suggestion-list")).not.toBeInTheDocument();
  });

  it("should not visible with showSuggestions=false", () => {
    render(<SuggestionsList suggestions={dataSuggestions} showSuggestions={false} />);
    expect(screen.queryByTestId("suggestion-list")).not.toBeInTheDocument();
  });
});
