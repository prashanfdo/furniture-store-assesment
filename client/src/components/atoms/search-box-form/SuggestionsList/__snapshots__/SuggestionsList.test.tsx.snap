// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionsList should render 1`] = `
<div>
  <ul
    class="fixed left-0 w-full pb-2 mt-2 bg-white rounded-md suggestion-list md:absolute top-16 md:top-full drop-shadow-md md:drop-shadow-lg"
    data-testid="suggestion-list"
  >
    <li>
      <a
        class="block px-4 py-2 decoration-transparent"
        href="/"
      >
        test 1
      </a>
    </li>
    <li>
      <a
        class="block px-4 py-2 decoration-transparent"
        href="/"
      >
        test 2
      </a>
    </li>
    <li>
      <a
        class="block px-4 py-2 decoration-transparent"
        href="/"
      >
        test 3
      </a>
    </li>
  </ul>
</div>
`;
